src/celengine/name.h

```diff
@@ -1,58 +1,140 @@
-//
-// C++ Interface: name
-//
-// Description:
-//
-//
-// Author: Toti <root@totibox>, (C) 2005
-//
-// Copyright: See COPYING file that comes with this distribution
-//
-//
 
 #pragma once
 
-#include <string>
-#include <iostream>
+#include <memory>
+#include <unordered_set>
+#include <unordered_map>
+#include <set>
 #include <map>
-#include <vector>
-#include <celutil/debug.h>
+#include <string>
 #include <celutil/util.h>
 #include <celutil/utf8.h>
 
-// TODO: this can be "detemplatized" by creating e.g. a global-scope enum InvalidCatalogNumber since there
-// lies the one and only need for type genericity.
-class NameDatabase
-{
- public:
-    typedef std::map<std::string, uint32_t, CompareIgnoringCasePredicate> NameIndex;
-    typedef std::multimap<uint32_t, std::string> NumberIndex;
-    enum {
-        InvalidCatalogNumber = 0xffffffff
-    };
+class AstroObject;
 
- public:
-    NameDatabase() {};
+class Name
+{
+protected:
+    std::shared_ptr<std::string> m_ptr { nullptr };
+    size_t m_hash { 0 };
+    void makeHash();
+    const static std::string m_empty;
+public:
+    Name(const char *s) { m_ptr = std::make_shared<std::string>(s); makeHash(); }
+    Name(const std::string& s) { m_ptr = std::make_shared<std::string>(s); makeHash(); }
```

Comment:
&s?

Reply:
More general questions: 1) why do want a sharable name? 2) why are trying to own a pointer to an object you don't own?

Reply:
1. On certain stage of developement I was afraid of number of string indexes. So I started looking for a method to save memory.
2. I don't understand what you mean. Probably because it's my first ever attempt to deal with shared_ptr. :-)